cdap-watchdog/src/main/java/io/cdap/cdap/metrics/collect/OTelMetricsCollectionService.java

```diff
@@ -0,0 +1,195 @@
+                return CacheBuilder.newBuilder().expireAfterAccess(CACHE_EXPIRE_MINUTES, TimeUnit.MINUTES).
+                        build(new CacheLoader<String, Pair<Attributes, Object>>() {
+                          @Override
+                          public Pair<Attributes, Object> load(String metricName) throws Exception {
+                            AttributesBuilder builder = Attributes.builder();
+                            tags.forEach((s1, s2) -> {
+                              builder.put(s1, s2);
+                            });
+                            return new Pair<>(builder.build(), null);
+                          }
+                        });
+              }
+            });
+  }
+
+  @Override
+  protected void run() throws Exception {
+    // TODO flush metrics on shutdown
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    // Flush the metrics when shutting down.
+    cloudMonitoringExporter.shutdown();
+  }
+
+  @Override
+  public MetricsContext getContext(Map<String, String> tags) {
+
+    return collectors.getUnchecked(tags);
+  }
+
+  private CacheLoader<Map<String, String>, MetricsContext> createCollectorLoader() {
+    return new CacheLoader<Map<String, String>, MetricsContext>() {
+      @Override
+      public MetricsContext load(final Map<String, String> collectorKey) throws Exception {
+        return new OTelMetricsCollectionService.OtelMetricsContextImpl(collectorKey);
+      }
+    };
+  }
+
+  private final class OtelMetricsContextImpl implements MetricsContext {
+    // TODO Decide Resource based on context.
+    // TODO once MeterProvider and meter creation should happen here based on metrics context
+
+    private final Map<String, String> tags;
+    private OtelMetricsContextImpl(final Map<String, String> tags) {
+      this.tags = ImmutableMap.copyOf(tags);
+    }
+
+    @Override
+    public void increment(String metricName, long value) {
+      Pair<Attributes, Object> counterPair = emitters.getUnchecked(tags).getUnchecked(metricName);
+      LongCounter counter;
+      if (counterPair.getValue() == null) {
+        // TODO: handle race of setting the counter
+        counter = meter.counterBuilder(metricName).setUnit("1").build();
+        emitters.getUnchecked(tags).put(metricName, new Pair<>(counterPair.getKey(), counter));
+      } else {
+        // TODO: check if counterPair.getValue() is LongCounter
+        counter = (LongCounter) counterPair.getValue();
+      }
+      counter.add(value, counterPair.getKey());
+    }
+
+    @Override
+    public void gauge(String metricName, long value) {
+      // TODO: there is no LongGauge in Otel?
+      Pair<Attributes, Object> gaugePair = emitters.getUnchecked(tags).getUnchecked(metricName);
+      ObservableDoubleGauge gauge;
+      if (gaugePair.getValue() == null) {
+        gauge = meter.gaugeBuilder(metricName).setUnit("1").buildWithCallback(null);
```

Comment:
This isn't how gauges work in OTEL.

In OTEL, a  Gauge needs to be callback-based and pull the current value (e.g. from an AtomicLong).

If you want to use a "last-value" aggregation of synchronously collected data here, you want to use a Histogram, and then configure a view on the SDK to convert it to a Gauge.

Reply:
Another option here, since you're already storing state, is to record the value you receive in the metric, and register a *single* callback to read that value from the table.